Added FulfillmentServiceService

davron12 opened the pull request:
Adds support for Shopify's FulfillmentService API

nozzlegear commented on ShopifySharp/Filters/ListFilter.cs:
```diff
@@ -40,5 +40,13 @@ public class ListFilter : CountFilter
+        /// Any valid values are null and "all"
+        /// </summary>
+        [JsonProperty("scope")]
+        public string Scope { get; set; }
```
Since the `ListFilter` class is used and inherited in several different places, could you create a new class named `FulfillmentServiceListFilter` that inherits from this `ListFilter` and adds the `Scope` property? That way we don't have the scope property showing up for unrelated services that also use this `ListFilter`.

It would look something like this:

```cs
public class FulfillmentServicesFilter : ListFilter 
{
    /// summary here
    [JsonProperty("scope")]
    public string Scope { get; set; }
}
```

davron12 replied:
Yes, only scope and none of the other parameters. I'll make it a parameter for the method then.

nozzlegear commented on ShopifySharp/Services/FulfillmentService/FulfillmentServiceService.cs:
```diff
@@ -0,0 +1,106 @@
+        /// </summary>
+        /// <param name="options">Options for filtering the list. Set scope to all to retrieve all of the store's fulfillment services</param>
+        /// <returns>The list of fulfillment services matching the filter.</returns>
+        public virtual async Task<IEnumerable<FulfillmentServiceEntity>> ListAsync(ListFilter filter = null)
```
Change `ListFilter filter = null` to `FulfillmentServicesFilter filter = null` once that class has been created.

nozzlegear commented on ShopifySharp/Entities/FulfillmentServiceEntity.cs:
```diff
@@ -0,0 +1,88 @@
+    /// <summary>
+    /// An object representing a Shopify fulfillment service.
+    /// </summary>
+    public class FulfillmentServiceEntity : ShopifyObject
```
Rename this class to `FulfillmentService` rather than `FulfillmentServiceEntity`.

nozzlegear replied:
Also the file should be `FulfillmentService.cs` once renamed, sorry!

davron12 replied:
That's originally what I wanted to name it, but because Entities and Services aren't in their own namespace, there will be a conflict with the Service already named FulfillmentService.

davron12 replied:
Yeah, I didn't like adding the suffix either, but didn't see any way around it.

nozzlegear commented on ShopifySharp/Services/FulfillmentService/FulfillmentServiceService.cs:
```diff
@@ -0,0 +1,106 @@
+    /// <summary>
+    /// A service for manipulating Shopify fulfillments.
+    /// </summary>
+    public class FulfillmentServiceService : ShopifyService
```
Just curious, I haven't looked at the docs yet. Does this endpoint support counting the Fulfillment Services?

davron12 replied:
No, there's no counting documented. I'm assuming most stores don't have a ton of fulfillment services registered, so there wasn't a reason to implement a count function.

davron12 commented:
I've made all the changes you requested, also fixed the spelling of Fulfillment in Fulfillment_Tests.cs, and added the OpenAsync method to the FulfillmentService.cs.

nozzlegear commented:
Published in 4.14.0, thank you very much!
